Welcome to the GitHub repository for the manuscript, "Genome-wide identification and annotation of multi-nucleotide variants in human reveal widespread functional differences compared to single nucleotide variants."

Here is the analysis code for each part of the manuscript. Each part has an R script `plot.r` for the figures that appear in the manuscript.

# Software + Comparison

# The largest database: Collected xx data to identify how many MNVs, overlap of different data, also includes

    Expanded MNV types, identified a large number of multi-joint MNVs
    Increased how many MNVs compared to gnomAD (how many multi-joint), a Venn diagram and gnomAD difference
    A database structure diagram, built an online database platform
    Corrected how many MNV frequency errors
    # Impact of MNVs on population stratification
    Compared MNV differences in different populations. 【What dimensions to compare】
    Use common SNVs and MNVs for population stratification. 【Check clustering effect】

# MNV base mutation pattern analysis

# MNVAnno corrected a large number of SNV annotation errors

    Compared genome-wide MNV distribution
    Corrected how many amino acid errors, LOF

# ncRNA

    # Impact on per-mRNA and lncRNA secondary structure
    # Impact on miRBS
    # If falling in seed region, the impact

# TFBS

    # How many TFBS errors,

# Functional application of MNV in cancer
