{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f8fa3f4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "efd99645",
   "metadata": {},
   "outputs": [],
   "source": [
    "### 012\n",
    "cancerList = [\"01.Breast\",\"02.Prostate\",\"03.Melanoma\",\"04.Colon\",\"05.Lung\",\"06.Cervical\",\"07.Uterine\",\"08.Bladder\",\"09.Rectal\",\"10.Ovary\",\"11.Kidney\"]\n",
    "\n",
    "for cancer in cancerList:\n",
    "    fo = open(\"/home/luohh/MNVAnalysis/01.GWAS/02.Cancer/01.Data/02.PhenoData/\"+cancer+\"/clinic.geno.012.tsv\",\"w\")\n",
    "\n",
    "    clinicDict = {}\n",
    "    with open(\"/home/luohh/MNVAnalysis/01.GWAS/02.Cancer/01.Data/02.PhenoData/\"+cancer+\"/clinic.tsv\",\"r\") as fi:\n",
    "        header = fi.readline().split()\n",
    "        fo.write(\"\\t\".join(header+[\"MNVID\",\"genotype\"])+\"\\n\")\n",
    "        for line in fi:\n",
    "            data = line.strip().split()\n",
    "            clinicDict[data[0]] = data\n",
    "    fi.close()\n",
    "\n",
    "    with open(\"/home/luohh/MNVAnalysis/01.GWAS/02.Cancer/01.Data/04.SurvivalMNVsGenoData/\"+cancer+\"/survival.vcf\",\"r\") as fi:\n",
    "        for line in fi:\n",
    "            if line[:2] == \"##\":\n",
    "                continue\n",
    "            elif line[:4] == \"#CHR\":\n",
    "                sampleIDList = line.strip().split()[9:]\n",
    "            else:\n",
    "                data = line.strip().split()\n",
    "                genoData = data[9:]\n",
    "                for i in range(len(sampleIDList)):\n",
    "                    if sampleIDList[i] in clinicDict:\n",
    "                        if genoData[i] == \"0/0\":\n",
    "                            fo.write(\"\\t\".join(clinicDict[sampleIDList[i]]+[data[2],\"0\"])+\"\\n\")\n",
    "                        elif genoData[i] == \"0/1\":\n",
    "                            fo.write(\"\\t\".join(clinicDict[sampleIDList[i]]+[data[2],\"1\"])+\"\\n\")\n",
    "                        elif genoData[i] == \"1/0\":\n",
    "                            fo.write(\"\\t\".join(clinicDict[sampleIDList[i]]+[data[2],\"1\"])+\"\\n\")\n",
    "                        elif genoData[i] == \"1/1\":\n",
    "                            fo.write(\"\\t\".join(clinicDict[sampleIDList[i]]+[data[2],\"2\"])+\"\\n\")\n",
    "                        else:\n",
    "                            fo.write(\"\\t\".join(clinicDict[sampleIDList[i]]+[data[2],\"NA\"])+\"\\n\")\n",
    "    fi.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72ec5685",
   "metadata": {},
   "outputs": [],
   "source": [
    "### 02\n",
    "cancerList = [\"01.Breast\",\"02.Prostate\",\"03.Melanoma\",\"04.Colon\",\"05.Lung\",\"06.Cervical\",\"07.Uterine\",\"08.Bladder\",\"09.Rectal\",\"10.Ovary\",\"11.Kidney\"]\n",
    "\n",
    "for cancer in cancerList:\n",
    "    fo = open(\"/home/luohh/MNVAnalysis/01.GWAS/02.Cancer/01.Data/02.PhenoData/\"+cancer+\"/clinic.geno.02.tsv\",\"w\")\n",
    "\n",
    "    clinicDict = {}\n",
    "    with open(\"/home/luohh/MNVAnalysis/01.GWAS/02.Cancer/01.Data/02.PhenoData/\"+cancer+\"/clinic.tsv\",\"r\") as fi:\n",
    "        header = fi.readline().split()\n",
    "        fo.write(\"\\t\".join(header+[\"MNVID\",\"genotype\"])+\"\\n\")\n",
    "        for line in fi:\n",
    "            data = line.strip().split()\n",
    "            clinicDict[data[0]] = data\n",
    "    fi.close()\n",
    "\n",
    "    with open(\"/home/luohh/MNVAnalysis/01.GWAS/02.Cancer/01.Data/04.SurvivalMNVsGenoData/\"+cancer+\"/survival.vcf\",\"r\") as fi:\n",
    "        for line in fi:\n",
    "            if line[:2] == \"##\":\n",
    "                continue\n",
    "            elif line[:4] == \"#CHR\":\n",
    "                sampleIDList = line.strip().split()[9:]\n",
    "            else:\n",
    "                data = line.strip().split()\n",
    "                genoData = data[9:]\n",
    "                for i in range(len(sampleIDList)):\n",
    "                    if sampleIDList[i] in clinicDict:\n",
    "                        if genoData[i] == \"0/0\":\n",
    "                            fo.write(\"\\t\".join(clinicDict[sampleIDList[i]]+[data[2],\"0\"])+\"\\n\")\n",
    "                        elif genoData[i] == \"0/1\":\n",
    "                            fo.write(\"\\t\".join(clinicDict[sampleIDList[i]]+[data[2],\"0\"])+\"\\n\")\n",
    "                        elif genoData[i] == \"1/0\":\n",
    "                            fo.write(\"\\t\".join(clinicDict[sampleIDList[i]]+[data[2],\"0\"])+\"\\n\")\n",
    "                        elif genoData[i] == \"1/1\":\n",
    "                            fo.write(\"\\t\".join(clinicDict[sampleIDList[i]]+[data[2],\"2\"])+\"\\n\")\n",
    "                        else:\n",
    "                            fo.write(\"\\t\".join(clinicDict[sampleIDList[i]]+[data[2],\"NA\"])+\"\\n\")\n",
    "    fi.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8df2513",
   "metadata": {},
   "outputs": [],
   "source": [
    "### 01\n",
    "cancerList = [\"01.Breast\",\"02.Prostate\",\"03.Melanoma\",\"04.Colon\",\"05.Lung\",\"06.Cervical\",\"07.Uterine\",\"08.Bladder\",\"09.Rectal\",\"10.Ovary\",\"11.Kidney\"]\n",
    "\n",
    "for cancer in cancerList:\n",
    "    fo = open(\"/home/luohh/MNVAnalysis/01.GWAS/02.Cancer/01.Data/02.PhenoData/\"+cancer+\"/clinic.geno.01.tsv\",\"w\")\n",
    "\n",
    "    clinicDict = {}\n",
    "    with open(\"/home/luohh/MNVAnalysis/01.GWAS/02.Cancer/01.Data/02.PhenoData/\"+cancer+\"/clinic.tsv\",\"r\") as fi:\n",
    "        header = fi.readline().split()\n",
    "        fo.write(\"\\t\".join(header+[\"MNVID\",\"genotype\"])+\"\\n\")\n",
    "        for line in fi:\n",
    "            data = line.strip().split()\n",
    "            clinicDict[data[0]] = data\n",
    "    fi.close()\n",
    "\n",
    "    with open(\"/home/luohh/MNVAnalysis/01.GWAS/02.Cancer/01.Data/04.SurvivalMNVsGenoData/\"+cancer+\"/survival.vcf\",\"r\") as fi:\n",
    "        for line in fi:\n",
    "            if line[:2] == \"##\":\n",
    "                continue\n",
    "            elif line[:4] == \"#CHR\":\n",
    "                sampleIDList = line.strip().split()[9:]\n",
    "            else:\n",
    "                data = line.strip().split()\n",
    "                genoData = data[9:]\n",
    "                for i in range(len(sampleIDList)):\n",
    "                    if sampleIDList[i] in clinicDict:\n",
    "                        if genoData[i] == \"0/0\":\n",
    "                            fo.write(\"\\t\".join(clinicDict[sampleIDList[i]]+[data[2],\"0\"])+\"\\n\")\n",
    "                        elif genoData[i] == \"0/1\":\n",
    "                            fo.write(\"\\t\".join(clinicDict[sampleIDList[i]]+[data[2],\"1\"])+\"\\n\")\n",
    "                        elif genoData[i] == \"1/0\":\n",
    "                            fo.write(\"\\t\".join(clinicDict[sampleIDList[i]]+[data[2],\"1\"])+\"\\n\")\n",
    "                        elif genoData[i] == \"1/1\":\n",
    "                            fo.write(\"\\t\".join(clinicDict[sampleIDList[i]]+[data[2],\"1\"])+\"\\n\")\n",
    "                        else:\n",
    "                            fo.write(\"\\t\".join(clinicDict[sampleIDList[i]]+[data[2],\"NA\"])+\"\\n\")\n",
    "    fi.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
