{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b9545681",
   "metadata": {},
   "outputs": [],
   "source": [
    "file= '/home/caow/03mnv/analyse3/02data/01origin/hg38mnv'\n",
    "url='/home/caow/03mnv/analyse3/02data/02annotation/'\n",
    "dataset='1000G,GTEx'\n",
    "maf_switch=0 # 如果是1的话，启用MAF过滤\n",
    "dataset=dataset.split(',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "03ff2e87",
   "metadata": {},
   "outputs": [],
   "source": [
    "mnv=[]\n",
    "with open(file) as f:\n",
    "    for i in f:\n",
    "        a=i.strip().split('\\t')\n",
    "        b=a[-1]\n",
    "        # if 'TCGA' in b or 'GTEx' in b or '1000G' in b or 'UKB20w' in b or 'UKB50w' in b:\n",
    "        #   mnv.append(a[2])\n",
    "        flag=0\n",
    "        for ii in dataset:\n",
    "            if ii in b:\n",
    "                flag=1\n",
    "                break\n",
    "        \n",
    "        if maf_switch==1:\n",
    "            if flag==1:\n",
    "                for ii in b.split('|'):\n",
    "                    if ii.startswith(dataset[0]):\n",
    "                        ac=int(ii.split('/')[0].split(':')[1])\n",
    "                        break\n",
    "                if ac>3: # 相当于maf=0.05%\n",
    "                    mnv.append(a[2])\n",
    "        else:\n",
    "            if flag==1:\n",
    "                mnv.append(a[2])\n",
    "\n",
    "mnv=set(mnv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "36a64fa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def createGeneAnno():\n",
    "    anno={}\n",
    "    anno['UTR5'] = 0\n",
    "    anno['exon'] = 0\n",
    "    anno['splice'] = 0\n",
    "    anno['intron'] = 0\n",
    "    anno['UTR3'] = 0\n",
    "    return anno\n",
    "anno=createGeneAnno()\n",
    "\n",
    "UTR5 = []\n",
    "exon = []\n",
    "splice = []\n",
    "intron = []\n",
    "UTR3 = []\n",
    "\n",
    "with open(url+'gene.txt') as f:\n",
    "    for i in f:\n",
    "        a=i.strip().split('\\t')\n",
    "        if a[-1] !='.' and a[2] in mnv:\n",
    "            if 'UTR5_variant' in a[-1]:\n",
    "                UTR5.append(a[2])\n",
    "            if 'exon' in a[-1]:\n",
    "                exon.append(a[2])\n",
    "            if 'splice_' in a[-1]:\n",
    "                splice.append(a[2])\n",
    "            if 'intron_variant' in a[-1]:\n",
    "                intron.append(a[2])\n",
    "            if 'UTR3_variant' in a[-1]:\n",
    "                UTR3.append(a[2])   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "55feb0dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "UTR5 = set(UTR5)\n",
    "exon = set(exon)\n",
    "splice = set(splice)\n",
    "intron = set(intron)\n",
    "UTR3 = set(UTR3)\n",
    "\n",
    "anno['UTR5']=len(UTR5)\n",
    "anno['exon']=len(exon)\n",
    "anno['splice']=len(splice)\n",
    "anno['intron']=len(intron)\n",
    "anno['UTR3']=len(UTR3)\n",
    "\n",
    "# lncRNA\n",
    "lnc=[]\n",
    "f=open(url+'lnc.txt')\n",
    "for i in f:\n",
    "    a=i.strip().split('\\t')\n",
    "    if a[2] in mnv and 'all lncRNA' in a[-1]:\n",
    "        lnc.append(a[2])\n",
    "f.close()\n",
    "lnc=set(lnc)\n",
    "anno['lncRNA']=len(lnc)\n",
    "\n",
    "# miR \n",
    "miR=[]\n",
    "f=open(url+'pre_miRNA.txt')\n",
    "for i in f:\n",
    "    a=i.strip().split('\\t')\n",
    "    if a[2] in mnv and 'all miRNA' in a[-1]:\n",
    "        miR.append(a[2])\n",
    "f.close()\n",
    "miR=set(miR)\n",
    "anno['miRNA']=len(miR)\n",
    "\n",
    "# circRNA\n",
    "circRNA=[]\n",
    "f=open(url+'circ.txt')\n",
    "for i in f:\n",
    "    a=i.strip().split('\\t')\n",
    "    if a[2] in mnv and 'all circRNA' in a[-1]:\n",
    "        circRNA.append(a[2])\n",
    "f.close()\n",
    "circRNA=set(circRNA)\n",
    "anno['circRNA']=len(circRNA)\n",
    "\n",
    "# piRNA\n",
    "piRNA=[]\n",
    "f=open(url+'piR.txt')\n",
    "for i in f:\n",
    "    a=i.strip().split('\\t')\n",
    "    if a[2] in mnv and 'all piRNA' in a[-1]:\n",
    "        piRNA.append(a[2])\n",
    "f.close()\n",
    "piRNA=set(piRNA)\n",
    "anno['piRNA']=len(piRNA)\n",
    "\n",
    "# ATAC\n",
    "ATAC=[]\n",
    "f=open(url+'atac.txt')\n",
    "for i in f:\n",
    "    a=i.strip().split('\\t')\n",
    "    if a[2] in mnv and 'all ATAC' in a[-1]:\n",
    "        ATAC.append(a[2])\n",
    "f.close()\n",
    "ATAC=set(ATAC)\n",
    "anno['ATAC']=len(ATAC)\n",
    "\n",
    "# CE\n",
    "conserved_element=[]\n",
    "f=open(url+'CE.txt')\n",
    "for i in f:\n",
    "    a=i.strip().split('\\t')\n",
    "    if a[2] in mnv and 'all conserved_element' in a[-1]:\n",
    "        conserved_element.append(a[2])\n",
    "f.close()\n",
    "conserved_element=set(conserved_element)\n",
    "anno['conserved_element']=len(conserved_element)\n",
    "\n",
    "# enhancer\n",
    "enhancer=[]\n",
    "f=open(url+'enhancer.txt')\n",
    "for i in f:\n",
    "    a=i.strip().split('\\t')\n",
    "    if a[2] in mnv and 'all enhancer' in a[-1]:\n",
    "        enhancer.append(a[2])\n",
    "f.close()\n",
    "enhancer=set(enhancer)\n",
    "anno['enhancer']=len(enhancer)\n",
    "\n",
    "# miRBS\n",
    "miRBS=[]\n",
    "f=open(url+'miRBS.txt')\n",
    "for i in f:\n",
    "    a=i.strip().split('\\t')\n",
    "    if a[2] in mnv and 'all miRBS' in a[-1]:\n",
    "        miRBS.append(a[2])\n",
    "f.close()\n",
    "miRBS=set(miRBS)\n",
    "anno['miRBS']=len(miRBS)\n",
    "\n",
    "# TFBS\n",
    "TFBS=[]\n",
    "f=open(url+'TFBS.txt')\n",
    "for i in f:\n",
    "    a=i.strip().split('\\t')\n",
    "    if a[2] in mnv and 'all TFBS' in a[-1]:\n",
    "        TFBS.append(a[2])\n",
    "f.close()\n",
    "TFBS=set(TFBS)\n",
    "anno['TFBS']=len(TFBS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "46fb563a",
   "metadata": {},
   "outputs": [],
   "source": [
    "intergenic=mnv-UTR5-exon-splice-intron-UTR3-lnc-miR-circRNA-piRNA\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5cb3e8d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "# mnv: set of MNV ids (你前面已经算出来了)\n",
    "# url: annotation dir\n",
    "\n",
    "mnv2anno = defaultdict(set)\n",
    "\n",
    "def add_anno_from_file(path, anno_func):\n",
    "    \"\"\"\n",
    "    path: 注释文件路径\n",
    "    anno_func: 输入该行 split 后的数组 a，返回一个或多个注释字符串（list/tuple/set），或返回空表示不记\n",
    "    \"\"\"\n",
    "    with open(path) as f:\n",
    "        for line in f:\n",
    "            line = line.strip()\n",
    "            if not line:\n",
    "                continue\n",
    "            a = line.split('\\t')\n",
    "            if len(a) < 3:\n",
    "                continue\n",
    "            mnvid = a[2]\n",
    "            if mnvid not in mnv:\n",
    "                continue\n",
    "\n",
    "            annos = anno_func(a)\n",
    "            if not annos:\n",
    "                continue\n",
    "            if isinstance(annos, str):\n",
    "                mnv2anno[mnvid].add(annos)\n",
    "            else:\n",
    "                for x in annos:\n",
    "                    if x:\n",
    "                        mnv2anno[mnvid].add(x)\n",
    "\n",
    "# gene 区域：可能一条记录同时命中多个（按你原脚本的 if 逻辑）\n",
    "def gene_annos(a):\n",
    "    cons = a[-1]\n",
    "    if cons == '.':\n",
    "        return []\n",
    "    out = []\n",
    "    if 'UTR5_variant' in cons:\n",
    "        out.append('UTR5')\n",
    "    if 'exon' in cons:\n",
    "        out.append('exon')\n",
    "    if 'splice_' in cons:\n",
    "        out.append('splice')\n",
    "    if 'intron_variant' in cons:\n",
    "        out.append('intron')\n",
    "    if 'UTR3_variant' in cons:\n",
    "        out.append('UTR3')\n",
    "    return out\n",
    "\n",
    "add_anno_from_file(url + 'gene.txt', gene_annos)\n",
    "\n",
    "# 其他大类：按原逻辑，只要包含某个“all xxx”标签就打这个注释\n",
    "add_anno_from_file(url + 'lnc.txt',        lambda a: 'lncRNA' if 'all lncRNA' in a[-1] else [])\n",
    "add_anno_from_file(url + 'mature_miRNA.txt',  lambda a: 'mature_miRNA'  if 'all miRNA' in a[-1] else [])\n",
    "add_anno_from_file(url + 'circ.txt',       lambda a: 'circRNA' if 'all circRNA' in a[-1] else [])\n",
    "add_anno_from_file(url + 'piR.txt',        lambda a: 'piRNA' if 'all piRNA' in a[-1] else [])\n",
    "add_anno_from_file(url + 'atac.txt',       lambda a: 'ATAC' if 'all ATAC' in a[-1] else [])\n",
    "add_anno_from_file(url + 'CE.txt',         lambda a: 'conserved_element' if 'all conserved_element' in a[-1] else [])\n",
    "add_anno_from_file(url + 'enhancer.txt',   lambda a: 'enhancer' if 'all enhancer' in a[-1] else [])\n",
    "add_anno_from_file(url + 'miRBS.txt',      lambda a: 'miRBS' if 'all miRBS' in a[-1] else [])\n",
    "add_anno_from_file(url + 'TFBS.txt',       lambda a: 'TFBS' if 'all TFBS' in a[-1] else [])\n",
    "\n",
    "# intergenic：不在任何注释集合中的 mnv\n",
    "for mnvid in mnv:\n",
    "    if mnvid not in mnv2anno:\n",
    "        mnv2anno[mnvid].add('intergenic')\n",
    "\n",
    "# 输出 TSV：两列 MNVid / annotation（多注释用 ; 连接）\n",
    "out_tsv = '/home/caow/03mnv/analyse3/00addition/data/1000G_GTEx_annotation.tsv'\n",
    "with open(out_tsv, 'w') as out:\n",
    "    out.write(\"MNVid\\tannotation\\n\")\n",
    "    for mnvid in sorted(mnv2anno.keys()):\n",
    "        annostr = \";\".join(sorted(mnv2anno[mnvid]))\n",
    "        out.write(f\"{mnvid}\\t{annostr}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3cc48e28",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.02269008830602263"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "2429195/(63993320+43066422)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ebcb491",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py36",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
